CORDA-785: Add functions for constructing FlowLogicRef without the class

rnicoll opened the pull request:
Removes the need for schedulable flow states to hold a reference to the class.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FlowLogicRef.kt:
```diff
@@ -11,6 +11,10 @@ import net.corda.core.serialization.CordaSerializable
 @DoNotImplement
 interface FlowLogicRefFactory {
     fun create(flowClass: Class<out FlowLogic<*>>, vararg args: Any?): FlowLogicRef
+    fun create(flow: FlowNameAndArguments): FlowLogicRef
```
All these methods need docs.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FlowLogicRef.kt:
```diff
@@ -11,6 +11,10 @@ import net.corda.core.serialization.CordaSerializable
+    fun create(flow: FlowNameAndArguments): FlowLogicRef
+    fun verify(flowClass: Class<out FlowLogic<*>>, vararg args: Any?): FlowNameAndArguments
+
+    class FlowNameAndArguments(val flowClass: String, val args: List<Pair<String, Any?>>)
```
Let's not use `Pair` for the public API as it's not very useful in Java. Looking at the code below is the String type the param name of the flow c'tor? Looks like it's only used for Kotlin?

rnicoll replied:
Unfortunately the `Pair` has to be exposed so that `FlowLogicRefFactoryImpl` can get to it, but isn't really intended for general use. There's no harm in it being visible, but it isn't the most useful data type I know.

shamsasari replied:
Am I correct in understanding that `Pair.first` are unique, and so this can be a Map?

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FlowLogicRef.kt:
```diff
@@ -11,6 +11,10 @@ import net.corda.core.serialization.CordaSerializable
+    fun create(flow: FlowNameAndArguments): FlowLogicRef
+    fun verify(flowClass: Class<out FlowLogic<*>>, vararg args: Any?): FlowNameAndArguments
+
+    class FlowNameAndArguments(val flowClass: String, val args: List<Pair<String, Any?>>)
```
Rename param to `flowClassName`

mkit replied:
Can it be a data class?

shamsasari commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowLogicRefFactoryImpl.kt:
```diff
@@ -35,42 +35,31 @@ object FlowLogicRefFactoryImpl : SingletonSerializeAsToken(), FlowLogicRefFactor
 
+    /**
+     * Construct a FlowLogicRef, intended for cases where the flow logic class is available.
+     */
```
This class is implementation detail and so this doc is in the wrong place. It should be in FlowLogicRefFactory

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FlowLogicRef.kt:
```diff
@@ -11,6 +11,10 @@ import net.corda.core.serialization.CordaSerializable
 @DoNotImplement
 interface FlowLogicRefFactory {
     fun create(flowClass: Class<out FlowLogic<*>>, vararg args: Any?): FlowLogicRef
```
Based on my reading of CORDA-785, should method be deprecated?

mkit commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowLogicRefFactoryImpl.kt:
```diff
@@ -138,4 +127,43 @@ object FlowLogicRefFactoryImpl : SingletonSerializeAsToken(), FlowLogicRefFactor
+        return verifyForRPC(flowClass, *args)
+    }
+
+    fun verifyForRPC(flowClass: Class<out FlowLogic<*>>, vararg args: Any?): FlowLogicRefFactory.FlowNameAndArguments {
```
Missing docs.

rnicoll commented:
Switched IRS to using this, and fixed up issues arising (turns out we can't actually pre-validate these, realistically, and instead we just pass in the class name).

rnicoll commented:
Rebased just to make sure there's no weird problems sneaking in. Any chance of a re-review @shamsasari ?

shamsasari commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowLogicRefFactoryImpl.kt:
```diff
@@ -42,6 +42,15 @@ object FlowLogicRefFactoryImpl : SingletonSerializeAsToken(), FlowLogicRefFactor
     }
 
+    override fun create(flowClassName: String, vararg args: Any?): FlowLogicRef {
+        val flowClass = Class.forName(flowClassName) as? Class<FlowLogic<*>>
```
Use `classloader` instead of whatever `forName` uses. Also I'm not sure why it's a `var`. See if you can make it into a `val`.

shamsasari replied:
You need to use `Class.isAssignableFrom` or `asSubclass` to correctly determine the class type.

shamsasari commented on core/src/main/kotlin/net/corda/core/flows/FlowLogicRef.kt:
```diff
@@ -10,7 +10,16 @@ import net.corda.core.serialization.CordaSerializable
+     * Construct a FlowLogicRef. This is intended for cases where the calling code has the relevant class already
+     * and can provide it directly.
+     */
     fun create(flowClass: Class<out FlowLogic<*>>, vararg args: Any?): FlowLogicRef
```
Shouldn't this method be deprecated as it just encourages the situation we're trying to avoid?

rnicoll commented:
@rick-r3 Any comments before I merge this in?
